static/js/biomedical/protein/protein_protein_interaction_graph.tsx

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Protein-protein interaction graph
+ */
+
+import axios from "axios";
+import _ from "lodash";
+import React from "react";
+
+import { drawProteinInteractionGraph } from "./chart";
+import { bioDcid, MultiLevelInteractionGraphData } from "./types";
+
+interface InteractionGraphProps {
+  centerProteinDcid: bioDcid;
+}
+
+interface InteractionGraphState {
+  depth: number;
+  graphData: MultiLevelInteractionGraphData;
+  numInteractions: number;
+  scoreThreshold: number;
+}
+
+const CHART_ID = "protein-interaction-graph";
+
+const DEFAULTS = {
+  DEPTH: 2,
+  MAX_INTERACTIONS: 4,
+  MISSING_SCORE_FILLER: -1,
+  SCORE_THRESHOLD: 0.4,
+};
+
+export class ProteinProteinInteractionGraph extends React.Component<
+  InteractionGraphProps,
+  InteractionGraphState
+> {
+  constructor(props: InteractionGraphProps) {
+    super(props);
+    this.state = {
+      depth: DEFAULTS.DEPTH,
+      graphData: null,
+      scoreThreshold: DEFAULTS.SCORE_THRESHOLD,
+      numInteractions: DEFAULTS.MAX_INTERACTIONS,
+    };
+  }
+
+  /**
+   * Perform BFS and (re)draw graph
+   */
+  componentDidUpdate(prevProps: InteractionGraphProps): void {
+    // takes two calls to this method to draw the graph.
+    //  1) first call retrieves graph data from flask endpoint
+    //  2) second call draws the graph
+
+    // this branch executes on first call to this method
+    if (prevProps !== this.props) {
+      axios
```

Comment:
- This also needs to be done in componentDidMount, otherwise on initial render of this component, data won't be fetched. 
- And since you'll be doing the same data fetch in two spots, should create a new private function in this component to do the data fetch.
- I think you also want to refetch data whenever state (except for graphData) changes? so maybe `if (!_.isEmpty(this.state.graphData) && !_.isEqual(prevState.graphData, this.state.graphData)) {draw and then return}` otherwise fetch data.

Reply:
Ok will do! Currently the graph still shows up initially in a hacky way because the parent component rerenders after the page request comes back --> rerenders the children and calls componentDidUpdate but your way is much more robust.

Reply:
ah, ok, then maybe also add a check that one of either props or state has changed before running the code in componentDidUpdate

Reply:
Oh ok will do!

Reply:
Ok I think fiddled a bit more and I think that for the original code I just got lucky in that I was using `prevProps !== this.props` previously instead of `_.isEqual(prevProps, this.props)`.  

I tried logging the following in the original version of the code:
```
  componentDidUpdate(prevProps: InteractionGraphProps): void {
    console.log('props', _.cloneDeep(prevProps), _.cloneDeep(this.props))
    console.log('1', _.isEqual(prevProps, this.props))
    console.log('2', prevProps === this.props)
    if (prevProps !== this.props) {
        //make request
            .then((resp) => {
              this.setState({ graphData: resp.data });
            });
        return;
    }
    // draw graph
```
and got
```
props {centerProteinDcid: 'bio/P53_HUMAN'} {centerProteinDcid: 'bio/P53_HUMAN'}
1 true
2 false
props {centerProteinDcid: 'bio/P53_HUMAN'} {centerProteinDcid: 'bio/P53_HUMAN'}
1 true
2 true
```
(so 1st call happens when Suhana's page request comes back, I guess a new props object is constructed so `prevProps !== this.props` is true, and then we update the state and get a 2nd call with the same props.)

I think basically we want a way to avoid calling `fetchData` twice (this can happen if we make the request in `componentDidMount` but it doesn't resolve by the time of the 1st call to `componentDidUpdate` from Suhana's page request coming back, so `state.graphData` is empty and we `fetchData` again).  Perhaps we can leave the first `fetchData` to componentDidMount, and in `componentDidUpdate` first check if `_.isEmpty(state.graphData)` and then return, and then check if the user input parts of the state are all the same as previously, and if so draw, and otherwise `fetchData`?

Reply:
Oops I think another problematic case is that if this component's `fetchData` arrives before Suhana's, then we draw the component twice.

Reply:
Ok I fixed this by adding a `drawn` property to the state.  So drawn is false initially and also set to false whenever we've fetched new data, and set to true when we call the draw function.

Reply:
I don't think the drawn flag is needed .. couldn't we just do this in the componentDidUpdate?
```
if (_.isEqual(prevProps, this.props) && _.isEqual(prevState, this.state) {
return
}
(!_.isEmpty(this.state.graphData) && !_.isEqual(prevState.graphData, this.state.graphData)) {
draw and then return
}
fetch data
```

Reply:
Oh! yes sorry this works thanks for writing it out!